include/sparrow/data_type.hpp

```diff
@@ -38,9 +74,117 @@ namespace sparrow
         // Variable-length bytes (no guarantee of UTF8-ness)
         BINARY,
         // Fixed-size binary. Each value occupies the same number of bytes
-        FIXED_SIZE_BINARY
+        FIXED_SIZE_BINARY,
     };
 
+    /// C++ types value representation types matching Arrow types.
+    // NOTE: this needs to be in sync-order with `data_type`
+    using all_base_types_t = mpl::typelist
+    <
+        std::nullopt_t  // REVIEW: not sure about if we need to have this one? for representing NA? is this the right type?
+    ,   bool
+    ,   std::uint8_t
+    ,   std::int8_t
+    ,   std::uint16_t
+    ,   std::int16_t
+    ,   std::uint32_t
+    ,   std::int32_t
+    ,   std::uint64_t
+    ,   std::int64_t
+    ,   float16_t
+    ,   float32_t
+    ,   float64_t
+    ,   std::string
+    //,   std::vector<std::byte> // REVIEW should this be uint8_t? char? buffer<unit8_t>?
```

Comment:
The main "classic" issue is that because `std::byte` has a strong type, not implicitly converted to other similar types, it's not automatically compatible with functions working with `char` or char-like types, in particular when interfacing with C. Lots of explicit casts ensures. Not a big problem if you like correctness but it does obfuscate the code. It's the main criticism I'm aware of.

In this library, I'm not sure what would be best, but the C api kinds of forces us with at least some of the types?

> Also, if we use uint8_t we must:

hmm actually shouldnt we always check that? I'll add it.

Reply:
The C Api uses `void*` for the buffers, so we'll have to cast anyway. I would actually be in favor of replacing the `uint8_t` with `byte` in the array_data buffers.

Reply:
Ok I went with an alias `byte_t` so that we can change that easilly, and added `vector<byte_t>` and I suppose we'll see how it goes from here.